example/lib/sum_types.dart

```diff
@@ -63,101 +63,164 @@ void main() {
+  static Something<T> address<T>(Optional<ty.Address> x) {
     return _Something(address: x);
   }
-  static Something something(Something x) {
+  static Something<T> something<T>(Something x) {
     return _Something(something: x);
   }
+  static Something<T> param<T>(T x) {
+    return _Something(param: x);
+  }
 }
 
-abstract class _SomethingBase {
+abstract class _SomethingBase<T> {
+  Optional<Unit> get nothing;
+  Optional<User> get user;
+  Optional<Optional<ty.Address>> get address;
+  Optional<Something> get something;
+  Optional<T> get param;
+
   __T iswitch<__T>({
     @required __T Function() nothing,
     @required __T Function(User) user,
     @required __T Function(Optional<ty.Address>) address,
     @required __T Function(Something) something,
+    @required __T Function(T) param,
   });
   __T iswitcho<__T>({
     __T Function() nothing,
     __T Function(User) user,
     __T Function(Optional<ty.Address>) address,
     __T Function(Something) something,
+    __T Function(T) param,
     @required __T Function() otherwise,
   });
 
   const _SomethingBase();
 }
 
-class _Something extends _SomethingBase with Something {
+class _Something<T> extends _SomethingBase<T> with Something<T> {
+
+  final Unit _nothing;
+  final User _user;
+  final Optional<ty.Address> _address;
+  final Something _something;
+  final T _param;
+
+  Optional<Unit> get nothing {
+    if (this._nothing != null) {
+      return Optional.of(this._nothing);
+    } else {
+      return Optional.absent();
+    }
+  }
+
+  Optional<User> get user {
+    if (this._user != null) {
+      return Optional.of(this._user);
+    } else {
+      return Optional.absent();
+    }
+  }
 
-  final Unit nothing;
-  final User user;
-  final Optional<ty.Address> address;
-  final Something something;
+  Optional<Optional<ty.Address>> get address {
+    if (this._address != null) {
+      return Optional.of(this._address);
+    } else {
+      return Optional.absent();
+    }
+  }
+  Optional<Something> get something {
+    if (this._something != null) {
+      return Optional.of(this._something);
+    } else {
+      return Optional.absent();
+    }
+  }
+  Optional<T> get param {
+    if (this._param != null) {
+      return Optional.of(this._param);
+    } else {
+      return Optional.absent();
+    }
+  }
 
-  const _Something({
-    this.nothing,
-    this.user,
-    this.address,
-    this.something
+  _Something({
+    Unit nothing,
+    User user,
+    Optional<ty.Address> address,
+    Something something,
+    T param,
   }) : assert(
          (nothing != null && user == null && address == null && something == null) ||
          (nothing == null && user != null && address == null && something == null) ||
          (nothing == null && user == null && address != null && something == null) ||
          (nothing == null && user == null && address == null && something != null)
-       );
+       ),
+       this._nothing = nothing,
```

Comment:
Warum nicht wie bisher? Der Construktor wird doch eh nur von generiertem Code aufgerufen. Dann kann auch das `const` wieder hin.

Reply:
What change do you mean?

Reply:
_Something({
        this._nothing

and so on

Reply:
Named optional parameters can't start with an underscore.

Reply:
Is there a reason this needs to be a named parameter in the first place?

Reply:
Named parameters given an additional sanity check for codegen. Also, I don't see a reason why we should optimize for code size here. The compiler removes all this stuff anyway.

Reply:
I would like to retain the `const` constructor as it allows the compiler to do more optimizations. If you omit that it is impossible to use these types in `const` contexts.

Reply:
You can't have a `const` constructor because of  https://github.com/dart-lang/sdk/issues/37810
It's not a question of named vs. unnamed arguments.

Reply:
Ah that's annoying. I thought that only applied if you declare those fields as variables instead of getters. In that case we can revisit this once that bug is fixed.